### PR TITLE
Feature/83 odometry

### LINKED ISSUE #39: Direction of motion estimation prototype
Prototype odometry module. 
Answer following points:
• Using camera specs
• Messages I/O
• Need for reference trajectory message

### LINKED ISSUE #40: Odometry module development
Odometry module as per prototype in #39

### PR DESCRIPTION
**Pull request by thibaudteil**
* **Tickets addressed:** bsk-000
* **Review:** By commit 
* **Merge strategy:** Merge (no squash)  

## Description
Add the visual odometry module, test, and documentation

## Verification
The module was unit tested and contains relevant assert statements

## Documentation
The module was documented with an RST file which includes the reference to the algorithm

## Future work
None

### CODE REVIEW

File: src/architecture/msgPayloadDefCpp/DirectionOfMotionMsgPayload.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+*/
+
+#ifndef DIRECTION_MOTION_MSG_H
+#define DIRECTION_MOTION_MSG_H
+
+#include "fswAlgorithms/imageProcessing/opticalFlow/opticalFlowDefinitions.h"
```
Comment (patkenneally) [inline src/architecture/msgPayloadDefCpp/DirectionOfMotionMsgPayload.h]: :large_orange_diamond:  Is this include and the below comment left over from a copy paste?

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.cpp [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+ MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+*/
+
+#include "fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h"
+#include "architecture/utilities/linearAlgebra.h"
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.cpp]: :small_red_triangle_down: Unused include.

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+/*! @brief Odometry module which reads key-point pairs and outputs a spacecraft direction of motion
+ */
+class VisualOdometry: public SysModel {
+public:
+    VisualOdometry();
+    ~VisualOdometry();
+
+    void Reset(uint64_t currentSimNanos) override;
+    void UpdateState(uint64_t currentSimNanos) override;
+
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h:46]: :large_orange_diamond: Extra new line.

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    BSKLogger bskLogger;                                                //!< -- BSK Logging
+
+    double errorTolerance = 1E-5;
+    double sigma_uv = 0.01;
+    double deltaKsi_tilde = 0.25;
+    int m_max = 10;
+
+private:
+    bool readMessages();
+    void writeMessages(Eigen::Vector3d sprime, Eigen::Matrix3d covar, uint64_t currentSimNanos);
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h]: :large_orange_diamond: Did you mean `sPrime`?

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    void computeX();
+    void computeRinv(Eigen::Vector3d sPrime);
+    void computeHTH();
+    Eigen::Vector3d svdLastColumn(Eigen::Matrix3d& A) const;
+    void computeKsis(const int i);
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h]: :large_orange_diamond: These don't need to be marked as const. The basic types (non-ref/pointer) will be copied into the function when called.

**Review by patkenneally — APPROVED** [review]

File: src/architecture/msgPayloadDefCpp/DirectionOfMotionMsgPayload.h (hunk 1 above)
Comment (thibaudteil) [reply src/architecture/msgPayloadDefCpp/DirectionOfMotionMsgPayload.h]: It is, deleted now

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.cpp (hunk 2 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.cpp]: Removed

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h (hunk 3 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h:46]: Removed

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h (hunk 4 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h]: Yes!

File: src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h (hunk 5 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/visualOdometry/visualOdometry.h]: removed

**Pull request merged**